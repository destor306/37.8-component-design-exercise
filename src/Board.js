import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";
import { copy } from "@testing-library/user-event/dist/cjs/clipboard/copy.js";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows =3, ncols=3, chanceLightStartsOn=.5 }) {
  
  const [board, setBoard] = useState(createBoard());

 
  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    // TODO: create array-of-arrays of true/false values
    return Array.from({length: ncols}).map(
      row => Array.from({length: nrows}).map(
        () =>Math.random() < chanceLightStartsOn
      )
    );
    
  }

  function hasWon() {
    // TODO: check the board in state to determine whether the player has won.
    return board.every(row => row.every(cell=> cell ===true))
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      // TODO: Make a (deep) copy of the oldBoard
      const copyBoard = oldBoard.map(row => [...row])
    
      // TODO: in the copy, flip this cell and the cells around it
      flipCell(y,x, copyBoard);
      flipCell(y+1,x, copyBoard)
      flipCell(y-1,x, copyBoard)
      flipCell(y,x+1, copyBoard)
      flipCell(y,x-1, copyBoard)

      // TODO: return the copy
      return copyBoard
    });
  }

  // if the game is won, just show a winning msg & render nothing else
  if(hasWon()){
    return <div>You WON!</div>
  }
  // TODO
  let tableboard = []

  const makeTable = (tableboard)=>{
    for(let i =0; i< ncols; i++){
      let row =[]
      for (let j =0;j<nrows; j++){
        let coord = `${i}-${j}`
        row.push(<Cell 
          key={coord}
          pos={coord} 
          isLit={board[i][j]} 
          flipCellsAroundMe={()=> flipCellsAround(coord)}/>)
      }
      tableboard.push(<tr key={i}>{row}</tr>)
    }
  }
 
  makeTable(tableboard);
  // TODO
  return(
    <table className="Board">
      <tbody>{tableboard}</tbody>
    </table>
  )
}

export default Board;
