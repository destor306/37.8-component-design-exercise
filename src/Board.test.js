import React from 'react';
import { render, fireEvent} from '@testing-library/react';
import Board from './Board';

const initialBoard = [
    [true, false, true],
    [false, true, false],
    [true, false, true]
  ];


it('should render', ()=>{
    render( <Board />)
})

// Mock the createBoard function to return a fixed initial board configuration
jest.mock('./Board', () => ({
    __esModule: true,
    default: jest.fn(({ nrows, ncols, chanceLightStartsOn }) => {
      // Create a fixed initial board configuration
      const initialBoard = [
        [true, false, true],
        [false, true, false],
        [true, false, true]
      ];
      return initialBoard;
    })
  }));

it('should match snapshot with fixed initial board configuration', () => {
    // Render the Board component with the fixed initial board configuration
    const { asFragment } = render(<Board  />);
    
    // Assert that the rendered output matches the stored snapshot
    expect(asFragment()).toMatchSnapshot();
  });


// it('should flip cells correctly on cell click', () => {
//     // Render the Board component
//     const { container  } = render(<Board />);

//     // Find the cell at row 0, column 0 (you can adjust the coordinates as needed)
//     const cellToClick = container.querySelector('Cell 0-0');
//     console.log(cellToClick);
//     // Simulate a click on the cell
//     fireEvent.click(cellToClick);

//     // Assert that the clicked cell has flipped
//     expect(cellToClick).toHaveClass('Cell-lit');

//     // Get all cells and check if the cells around the clicked cell have flipped
//     const adjacentCellPositions = [
//         '-1-0', '1-0', '0-1', '0-1' // Positions of adjacent cells
//     ];
//     adjacentCellPositions.forEach(pos => {
//         const adjacentCell = getByText(pos);
//         expect(adjacentCell).toHaveClass('Cell-lit');
//     });
// });