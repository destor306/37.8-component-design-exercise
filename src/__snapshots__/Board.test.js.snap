// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot with fixed initial board configuration 1`] = `<DocumentFragment />`;
